File: app/src/main/java/com/example/foolcardgame/domain/bot/BotAI.kt
package com.example.foolcardgame.domain.bot

import com.example.foolcardgame.domain.engine.Rules
import com.example.foolcardgame.domain.model.Card
import com.example.foolcardgame.domain.model.GamePhase
import com.example.foolcardgame.domain.model.GameState
import com.example.foolcardgame.domain.model.canAddMoreAttacks
import com.example.foolcardgame.domain.model.helperThrowerIds
import com.example.foolcardgame.domain.model.permissionsFor

/**
 * Simple deterministic bot: lowest legal card / obvious pass-bito-take.
 */
class BotAI {

    sealed class Action {
        data class Ready(val playerId: String) : Action()
        data class PlayCard(val playerId: String, val card: Card, val targetPairId: Int?) : Action()
        data class AddCard(val playerId: String, val card: Card) : Action()
        data class Pass(val playerId: String) : Action()
        data class Bito(val playerId: String) : Action()
    }

    /**
     * @param controlAllPlayers when true (functional tests), any player can be driven by AI.
     */
    fun chooseAction(state: GameState, controlAllPlayers: Boolean = false): Action? {
        fun controllable(isBot: Boolean) = controlAllPlayers || isBot

        when (state.phase) {
            GamePhase.LOBBY_WAITING -> {
                val bot = state.players.firstOrNull {
                    controllable(it.isBot) && !it.isReady
                } ?: return null
                return Action.Ready(bot.id)
            }
            GamePhase.FINISHED -> return null
            GamePhase.IN_PROGRESS -> Unit
        }

        // Mid-defense: any non-defender may throw in parallel.
        if (state.unbeatenPairs.isNotEmpty()) {
            chooseThrowIn(state, { controllable(it) }, includeAttacker = true)?.let { return it }
        }

        val defenderId = state.defenderId
        if (defenderId != null && state.unbeatenPairs.isNotEmpty()) {
            val defender = state.player(defenderId)
            if (defender != null && controllable(defender.isBot)) {
                return chooseDefenseOrTake(state, defenderId)
            }
            if (!controlAllPlayers) {
                // Human defending — only parallel throws (already tried) remain for bots.
                return null
            }
        }

        val attackerId = state.attackerId
        if (attackerId != null) {
            val attacker = state.player(attackerId)
            if (attacker != null && controllable(attacker.isBot) &&
                state.permissionsFor(attackerId).canBito
            ) {
                return Action.Bito(attackerId)
            }
        }

        // After attacker «Бито»: helpers throw if possible, else confirm.
        if (state.attackerBitoDeclared && state.allBeaten) {
            chooseThrowIn(state, { controllable(it) }, includeAttacker = false)?.let { return it }
            for (helperId in state.helperThrowerIds()) {
                if (helperId in state.passedPlayerIds) continue
                val helper = state.player(helperId) ?: continue
                if (!controllable(helper.isBot)) continue
                if (state.permissionsFor(helperId).canPass) {
                    return Action.Pass(helperId)
                }
            }
        }

        // All beaten, attacker not yet declared: helpers may still throw.
        if (state.allBeaten && !state.attackerBitoDeclared) {
            chooseThrowIn(state, { controllable(it) }, includeAttacker = false)?.let { return it }
        }

        if (!controlAllPlayers) {
            val actorId = state.currentPlayerId ?: return null
            val actor = state.player(actorId) ?: return null
            if (!controllable(actor.isBot)) return null
            return chooseActionForActor(state, actorId)
        }

        if (state.tablePairs.isEmpty() && attackerId != null) {
            val attacker = state.player(attackerId) ?: return null
            if (!controllable(attacker.isBot)) return null
            val card = attacker.hand.minWithOrNull(lowestFirst(state.trumpSuit)) ?: return null
            return Action.PlayCard(attackerId, card, targetPairId = null)
        }

        return null
    }

    private fun chooseDefenseOrTake(state: GameState, defenderId: String): Action {
        val trump = state.trumpSuit ?: return Action.Pass(defenderId)
        val unbeaten = state.unbeatenPairs.first()
        val beating = state.player(defenderId)?.hand
            ?.filter { Rules.beats(it, unbeaten.attack, trump) }
            ?.minWithOrNull(lowestFirst(trump))
        return if (beating != null) {
            Action.PlayCard(defenderId, beating, unbeaten.id)
        } else {
            Action.Pass(defenderId)
        }
    }

    private fun chooseThrowIn(
        state: GameState,
        controllable: (Boolean) -> Boolean,
        includeAttacker: Boolean,
    ): Action? {
        if (state.tablePairs.isEmpty() || !state.canAddMoreAttacks()) return null
        val candidates = state.players.filter { player ->
            controllable(player.isBot) &&
                player.id != state.defenderId &&
                player.id !in state.passedPlayerIds &&
                (includeAttacker || player.id != state.attackerId) &&
                !(state.attackerBitoDeclared && player.id == state.attackerId) &&
                player.hand.isNotEmpty()
        }
        for (player in candidates) {
            val throwCard = player.hand
                .filter { Rules.canThrow(it, state.tableRanks) }
                .minWithOrNull(lowestFirst(state.trumpSuit))
            if (throwCard != null) {
                return Action.AddCard(player.id, throwCard)
            }
        }
        return null
    }

    private fun chooseActionForActor(state: GameState, actorId: String): Action? {
        val perms = state.permissionsFor(actorId)
        val attackerId = state.attackerId

        if (actorId == state.defenderId && state.unbeatenPairs.isNotEmpty()) {
            return chooseDefenseOrTake(state, actorId)
        }

        if (perms.canBito) {
            return Action.Bito(actorId)
        }

        if (perms.canPass) {
            return Action.Pass(actorId)
        }

        if (state.tablePairs.isEmpty() && actorId == attackerId) {
            val card = state.player(actorId)?.hand
                ?.minWithOrNull(lowestFirst(state.trumpSuit))
                ?: return null
            return Action.PlayCard(actorId, card, targetPairId = null)
        }

        return null
    }

    private fun lowestFirst(trumpSuit: com.example.foolcardgame.domain.model.Suit?) =
        Comparator<Card> { a, b ->
            val aTrump = trumpSuit != null && a.suit == trumpSuit
            val bTrump = trumpSuit != null && b.suit == trumpSuit
            when {
                aTrump != bTrump -> if (aTrump) 1 else -1
                a.rank != b.rank -> a.rank.compareTo(b.rank)
                else -> a.suit.compareTo(b.suit)
            }
        }
}

File: server/src/internal/simclient/bot_ai.go
package simclient

import (
	"foolcardgame/server/internal/pb"
)

// chooseLegalAction выбирает легальный ход (как BotAI на Android: сначала отбивка, потом беру).
func chooseLegalAction(st *pb.GameState) *pb.ClientMessage {
	if st == nil || st.Phase != pb.GamePhase_IN_PROGRESS {
		return nil
	}

	trump := st.Trump
	localID := st.LocalPlayerId
	isDefender := st.DefenderId != nil && *st.DefenderId == localID
	isAttacker := st.AttackerId != nil && *st.AttackerId == localID

	var undefended *pb.TablePair
	for _, p := range st.TablePairs {
		if p != nil && p.Defense == nil {
			undefended = p
			break
		}
	}

	// Атакующий: «Бито», когда всё отбито.
	if st.CanBito {
		return &pb.ClientMessage{Payload: &pb.ClientMessage_Bito{Bito: &pb.Bito{}}}
	}

	// Защитник: сначала попытка отбить, иначе «беру».
	if isDefender && undefended != nil {
		beating := lowestBeating(st.LocalHand, undefended.Attack, trump)
		if beating != nil {
			pid := undefended.Id
			return &pb.ClientMessage{Payload: &pb.ClientMessage_PlayCard{PlayCard: &pb.PlayCard{
				Card: beating, TargetPairId: &pid,
			}}}
		}
		if st.CanTake {
			return &pb.ClientMessage{Payload: &pb.ClientMessage_Pass{Pass: &pb.Pass{}}}
		}
		return nil
	}

	// Помощник: подтверждение бито (Pass), если нельзя/нечего подкинуть.
	ranks := tableRanks(st)
	if !isDefender && len(st.TablePairs) > 0 {
		for _, c := range st.LocalHand {
			if c == nil {
				continue
			}
			if _, ok := ranks[c.Rank]; ok {
				return &pb.ClientMessage{Payload: &pb.ClientMessage_AddCard{AddCard: &pb.AddCard{Card: c}}}
			}
		}
	}
	if st.CanPass {
		return &pb.ClientMessage{Payload: &pb.ClientMessage_Pass{Pass: &pb.Pass{}}}
	}

	// Первая атака
	if isAttacker && len(st.TablePairs) == 0 && st.CurrentPlayerId != nil && *st.CurrentPlayerId == localID {
		if c := lowestCard(st.LocalHand, trump); c != nil {
			return &pb.ClientMessage{Payload: &pb.ClientMessage_PlayCard{PlayCard: &pb.PlayCard{Card: c}}}
		}
	}

	return nil
}

func lowestBeating(hand []*pb.Card, attack, trump *pb.Card) *pb.Card {
	var best *pb.Card
	for _, c := range hand {
		if c == nil || !beats(c, attack, trump) {
			continue
		}
		if best == nil || cardLess(c, best, trump) {
			best = c
		}
	}
	return best
}

func tableRanks(st *pb.GameState) map[pb.Rank]struct{} {
	m := make(map[pb.Rank]struct{})
	for _, p := range st.TablePairs {
		if p == nil || p.Attack == nil {
			continue
		}
		m[p.Attack.Rank] = struct{}{}
		if p.Defense != nil {
			m[p.Defense.Rank] = struct{}{}
		}
	}
	return m
}

func beats(defense, attack, trump *pb.Card) bool {
	if defense == nil || attack == nil {
		return false
	}
	if defense.Suit == attack.Suit {
		return defense.Rank > attack.Rank
	}
	if trump != nil && defense.Suit == trump.Suit && attack.Suit != trump.Suit {
		return true
	}
	if trump != nil && defense.Suit == trump.Suit && attack.Suit == trump.Suit {
		return defense.Rank > attack.Rank
	}
	return false
}

func lowestCard(hand []*pb.Card, trump *pb.Card) *pb.Card {
	var best *pb.Card
	for _, c := range hand {
		if c == nil {
			continue
		}
		if best == nil || cardLess(c, best, trump) {
			best = c
		}
	}
	return best
}

func cardLess(a, b, trump *pb.Card) bool {
	aTrump := trump != nil && a.Suit == trump.Suit
	bTrump := trump != nil && b.Suit == trump.Suit
	if aTrump != bTrump {
		return !aTrump && bTrump
	}
	return a.Rank < b.Rank
}
